lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+        """
+        return await self._broadcast(item, config, None)
+
+    async def _broadcast(
+        self,
+        item: BaseEvent,
+        config: Optional[BroadcastConfig],
+        id: Optional[RequestID],
+    ) -> None:
+        """
+        Broadcast an instance of :class:`~lahja.common.BaseEvent` on the event bus. Takes
+        an optional second parameter of :class:`~lahja.common.BroadcastConfig` to decide
+        where this event should be broadcasted to. By default, events are broadcasted across
+        all connected endpoints with their consuming call sites.
+        """
+        item.bind(self, id)
+
+        is_eligible = should_endpoint_receive_item(
+            item, config, self.name, self.get_subscribed_events()
+        )
+        is_internal = config is not None and config.internal
+
+        if is_eligible:
+            await self._process_item(item, config)
+
+        if is_internal:
+            # if the event is flagged as internal we exit early since it should
+            # not be broadcast beyond this endpoint
+            return
+
+        remotes_for_broadcast = tuple(
+            remote
+            for remote in self._connections
+            if should_endpoint_receive_item(
+                item, config, remote.name, remote.get_subscribed_events()
+            )
+        )
+
+        compressed_item = self._compress_event(item)
+        message = Broadcast(compressed_item, config)
+        for remote in remotes_for_broadcast:
+            try:
+                await remote.send_message(message)
+            except RemoteDisconnected as err:
+                self.logger.debug(
+                    "%s: dropping disconnected remote %s: %s", remote, err
+                )
+                await remote.stop()
+
+    def broadcast_nowait(
+        self, item: BaseEvent, config: Optional[BroadcastConfig] = None
+    ) -> None:
+        self._broadcast_send_channel.send_nowait((item, config))
+
+    TResponse = TypeVar("TResponse", bound=BaseEvent)
+
+    async def request(
+        self,
+        item: BaseRequestResponseEvent[TResponse],
+        config: Optional[BroadcastConfig] = None,
+    ) -> TResponse:
+        """
+        Broadcast an instance of
+        :class:`~lahja.common.BaseRequestResponseEvent` on the event bus and
+        immediately wait on an expected answer of type
+        :class:`~lahja.common.BaseEvent`. Optionally pass a second parameter of
+        :class:`~lahja.common.BroadcastConfig` to decide where the request
+        should be broadcasted to. By default, requests are broadcasted across
+        all connected endpoints with their consuming call sites.
+        """
+        request_id = next(self._get_request_id)
+
+        # Create an asynchronous generator that we use to pipe the result
+        send_channel, receive_channel = cast(
+            RequestResponseChannelPair, trio.open_memory_channel(0)
+        )
```

Comment:
I'm surprised that trio doesn't have a `Future`. A channel can do everything a future can do and more, sure, but this is so much more code than a future would have been!

Reply:
Yeah, I was surprised as well.  Supposedly it should be pretty easy to write one..